src/articles/articles.service.ts

```diff
@@ -224,4 +225,133 @@ export class ArticlesService {
 
     await this.prisma.comment.delete({ where: { id: commentId } });
   }
+
+  async getArticles(query: QueryArticleDto, currentUser?: User) {
+    const whereClause: Prisma.ArticleWhereInput = {};
+
+    if (query.tag) {
+      whereClause.tags = { some: { name: query.tag } };
+    }
+
+    if (query.author) {
+      whereClause.author = { username: query.author };
+    }
+
+    if (query.favorited) {
+      whereClause.favoritedBy = { some: { username: query.favorited } };
+    }
+
+    const [articles, articlesCount] = await Promise.all([
+      this.prisma.article.findMany({
+        where: whereClause,
+        skip: Number(query.offset) || 0,
+        take: Number(query.limit) || 20,
+        orderBy: { createdAt: 'desc' },
+        include: { author: true, tags: true, favoritedBy: true },
+      }),
+      this.prisma.article.count({ where: whereClause }),
+    ]);
+
+    let followingIds: number[] = [];
+    if (currentUser) {
+      followingIds =
+        (
+          await this.prisma.user.findUnique({
+            where: { id: currentUser.id },
+            select: { following: { select: { id: true } } },
+          })
+        )?.following.map((f) => f.id) ?? [];
+    }
+
+    const formattedArticles = await Promise.all(
+      articles.map((article) =>
+        this.buildArticleResponse(article, currentUser, followingIds),
+      ),
+    );
+
+    return {
+      articles: formattedArticles.map((a) => a.article),
+      articlesCount,
+    };
+  }
+
+  async getFeed(currentUser: User, query: QueryArticleDto) {
+    const followedUsers = await this.prisma.user.findUnique({
+      where: { id: currentUser.id },
+      select: { following: { select: { id: true } } },
+    });
+
+    const followedUserIds =
+      followedUsers?.following.map((user) => user.id) ?? [];
+
+    if (followedUserIds.length === 0) {
+      return { articles: [], articlesCount: 0 };
+    }
+
+    const [articles, articlesCount] = await Promise.all([
+      this.prisma.article.findMany({
+        where: { authorId: { in: followedUserIds } },
+        skip: Number(query.offset) ?? 0,
+        take: Number(query.limit) ?? 20,
+        orderBy: { createdAt: 'desc' },
+        include: { author: true, tags: true, favoritedBy: true },
+      }),
+      this.prisma.article.count({
+        where: { authorId: { in: followedUserIds } },
+      }),
+    ]);
+
+    const formattedArticles = await Promise.all(
+      articles.map((article) =>
+        this.buildArticleResponse(article, currentUser, followedUserIds),
+      ),
+    );
+
+    return {
+      articles: formattedArticles.map((a) => a.article),
+      articlesCount,
+    };
+  }
+
+  private async buildArticleResponse(
```

Comment:
```suggestion
  private buildArticleResponse(
```
hàm này ko cần async phải ko nhỉ?